### ISSUE TITLE
Unclear how to use merge with expression condition

### ISSUE THREAD

**Issue by koertkuipers** [delta-io/delta#419]
this used to work with older version of delta :
```
val data = Seq((1, 1, 1), (2, 2, 2)).toDF("part", "id", "x")
data.write
  .format("delta")
  .partitionBy("part")
  .save(tempPath)
val table = io.delta.tables.DeltaTable.forPath(spark, tempPath)
val tableDf = table.toDF
val data1  = Seq((2, 2, 4), (2, 3, 6), (3, 3, 9)).toDF("part", "id", "x")
table.merge(
  data1,
  tableDf("part") === data1("part") && tableDf("id") === data1("id")
)
  .whenMatched.updateAll
  .whenNotMatched.insertAll
  .execute()
```
now it produces error:
```
org.apache.spark.sql.AnalysisException: cannot resolve `part` in UPDATE clause given columns `_1`, `_2`, `_3`;
```
this is because something seems to have changed in table.toDF i think?
i can get it work again like this:
```
val data = Seq((1, 1, 1), (2, 2, 2)).toDF("part", "id", "x")
data.write
  .format("delta")
  .partitionBy("part")
  .save(tempPath)
val table = io.delta.tables.DeltaTable.forPath(spark, tempPath)
val tableDf = table.toDF
val data1  = Seq((2, 2, 4), (2, 3, 6), (3, 3, 9)).toDF("part", "id", "x")
table.alias("old").merge(
  data1.alias("new"),
  "old.part = new.part AND old.id = new.id"
)
  .whenMatched.updateAll
  .whenNotMatched.insertAll
  .execute()
```
so this relies on the merge condition being a string which is maybe a little less nice i think. does that mean there is no nice way anymore to have the merge condition be an expression (unambiguously referencing dataframes)?
thanks

**Comment by tdas** [comment]
Which versions are you talking about?

**Comment by koertkuipers** [comment]
its broken with master. it used to work with 0.4.x

**Comment by zsxwing** [comment]
@koertkuipers Thanks for reporting this and the great reproducer. Confirmed that this is broken in 0.6.0.

**Comment by tdas** [comment]
@koertkuipers we are fixing this, and we should make a 0.6.1 release soon with this fix.

**Comment by tdas** [comment]
Fixed in the following commits 
- master - 8da5df0421ab159d74ae5d65d5d1d9c03dc48b74 and 8ddb49ab86c7417938fde0bfb74dedeb7e6ae59d
- Branch 0.6 (to be released in 0.6.1) - 88ffb1d4069317131f5ba7c29894c4c82cb817d5 and 3aad9890e3e8596947db430e52e8142811b2bc59
